File: usda-nifa-gp-proposal/scripts/create_temperature_data.py
#!/usr/bin/env python3
"""
Create realistic daily temperature and photoperiod data for soybean trial locations.

This script generates climate-realistic data based on regional norms.
For actual NASA POWER data, run the R script with nasapower or apsimx packages.
"""

import csv
from datetime import datetime, timedelta
import math
import random

def calculate_photoperiod(doy, latitude_deg):
    """Calculate photoperiod (hours) for a given DOY and latitude (Spencer 1971)."""
    latitude_rad = math.radians(latitude_deg)
    b = (doy - 1) * 2 * math.pi / 365
    decl = (0.006918 - 0.399912 * math.cos(b) + 0.070257 * math.sin(b) -
            0.006758 * math.cos(2*b) + 0.000907 * math.sin(2*b) -
            0.00002 * math.cos(3*b) + 0.00029 * math.sin(3*b))
    cos_h = -math.tan(latitude_rad) * math.tan(decl)
    cos_h = max(-1, min(1, cos_h))
    h = math.acos(cos_h)
    return 24 * h / math.pi

# Regional climate parameters (based on NOAA data)
locations = {
    "Fayetteville, AR": {
        "lat": 36.0726,
        "lon": -94.1729,
        "seasonal_temp_mean": 24.8,  # May-Aug average
        "seasonal_temp_amplitude": 8.5,  # ±amplitude from mean
        "noise_std": 2.5  # daily variability
    },
    "Columbia, MO": {
        "lat": 38.2526,
        "lon": -92.2734,
        "seasonal_temp_mean": 23.2,  # May-Aug average (cooler than AR)
        "seasonal_temp_amplitude": 9.0,
        "noise_std": 2.8
    }
}

sowing_dates = {
    "Early (Apr 15)": {"date": "2023-04-15", "doy": 105},
    "Late (Jun 15)": {"date": "2023-06-15", "doy": 166}
}

all_data = []
random.seed(42)

print("Generating realistic daily temperature and photoperiod data...")
print("=" * 70)

for loc_name, loc_info in locations.items():
    for sow_name, sow_info in sowing_dates.items():
        print(f"\n{loc_name} - {sow_name}")

        start_dt = datetime.strptime(sow_info["date"], "%Y-%m-%d")
        end_dt = start_dt + timedelta(days=129)

        scenario_name = f"{loc_name} - {sow_name}"

        for day_offset in range(130):  # 130 days for 130-day crop cycle
            current_dt = start_dt + timedelta(days=day_offset)
            doy = current_dt.timetuple().tm_yday

            # Temperature model: seasonal curve + daily noise
            # Simple sine curve from April to September
            season_phase = (doy - 105) / 180 * math.pi  # 0 at DOY 105, pi at ~290
            if season_phase < 0:
                season_phase = 0
            elif season_phase > math.pi:
                season_phase = math.pi

            seasonal_mean = (loc_info["seasonal_temp_mean"] +
                           loc_info["seasonal_temp_amplitude"] *
                           math.sin(season_phase))
            daily_noise = random.gauss(0, loc_info["noise_std"])
            temp_mean = seasonal_mean + daily_noise

            # Diurnal temperature range: ~8-10°C for growing season
            diurnal_range = 9.0
            temp_min = temp_mean - diurnal_range / 2
            temp_max = temp_mean + diurnal_range / 2

            # Calculate photoperiod
            pp = calculate_photoperiod(doy, loc_info["lat"])

            all_data.append({
                "location": loc_name,
                "latitude": loc_info["lat"],
                "sowing_date": sow_name,
                "sowing_doy": sow_info["doy"],
                "calendar_date": current_dt.strftime("%Y-%m-%d"),
                "calendar_doy": doy,
                "day_in_season": day_offset + 1,
                "temp_mean": round(temp_mean, 1),
                "temp_min": round(temp_min, 1),
                "temp_max": round(temp_max, 1),
                "photoperiod_hours": round(pp, 2),
                "scenario": scenario_name
            })

        print(f"  ✓ Generated 130 days of data")

# Save to CSV
output_file = "temperature_and_photoperiod_data.csv"
with open(output_file, 'w', newline='') as f:
    writer = csv.DictWriter(f, fieldnames=all_data[0].keys())
    writer.writeheader()
    writer.writerows(all_data)

print("\n" + "=" * 70)
print(f"✓ Data saved to: {output_file}")
print(f"✓ Total records: {len(all_data)}")

# Summary statistics
print(f"\nTemperature range across all scenarios:")
temps = [float(d["temp_mean"]) for d in all_data]
print(f"  Min: {min(temps):.1f}°C")
print(f"  Max: {max(temps):.1f}°C")
print(f"  Mean: {sum(temps)/len(temps):.1f}°C")

print(f"\nPhotoperiod range across all scenarios:")
pps = [float(d["photoperiod_hours"]) for d in all_data]
print(f"  Min: {min(pps):.1f} hours")
print(f"  Max: {max(pps):.1f} hours")

# By scenario
scenarios = {}
for d in all_data:
    if d["scenario"] not in scenarios:
        scenarios[d["scenario"]] = {"temps": [], "pps": []}
    scenarios[d["scenario"]]["temps"].append(float(d["temp_mean"]))
    scenarios[d["scenario"]]["pps"].append(float(d["photoperiod_hours"]))

print(f"\nTemperature summary by scenario:")
for scenario in sorted(scenarios.keys()):
    temps = scenarios[scenario]["temps"]
    print(f"  {scenario}:")
    print(f"    Mean: {sum(temps)/len(temps):.1f}°C (range: {min(temps):.1f}–{max(temps):.1f}°C)")

print(f"\nPhotoperiod summary by scenario:")
for scenario in sorted(scenarios.keys()):
    pps = scenarios[scenario]["pps"]
    print(f"  {scenario}:")
    print(f"    Range: {min(pps):.1f}–{max(pps):.1f} hours")

print("\n" + "=" * 70)
print("NOTE: This is climate-realistic simulated data for visualization.")
print("      For actual NASA POWER data, use the R script:")
print("      rscript fetch_nasa_power_real_data.R")

File: usda-nifa-gp-proposal/scripts/create_temperature_data.R
#!/usr/bin/env Rscript
## ============================================================
## Generate climate-realistic daily temperature and photoperiod data
## for soybean trial locations (replacement for Python script)
## ============================================================

suppressPackageStartupMessages({
  library(dplyr)
  library(lubridate)
})

## ── Photoperiod calculation (Spencer 1971) ──────────────────
calc_photoperiod <- function(doy, latitude_deg) {
  latitude_rad <- latitude_deg * pi / 180
  b <- (doy - 1) * 2 * pi / 365
  decl <- 0.006918 - 0.399912 * cos(b) + 0.070257 * sin(b) -
          0.006758 * cos(2*b) + 0.000907 * sin(2*b) -
          0.00002 * cos(3*b) + 0.00029 * sin(3*b)
  cos_h <- -tan(latitude_rad) * tan(decl)
  cos_h <- pmax(-1, pmin(1, cos_h))
  h <- acos(cos_h)
  24 * h / pi
}

## ── Regional climate parameters ──────────────────────────────
locations <- tribble(
  ~location, ~lat, ~lon,
  "Fayetteville, AR", 36.0726, -94.1729,
  "Columbia, MO", 38.2526, -92.2734
)

location_params <- list(
  "Fayetteville, AR" = list(
    seasonal_temp_mean = 24.8,   # May-Aug average
    seasonal_temp_amplitude = 8.5,
    noise_std = 2.5
  ),
  "Columbia, MO" = list(
    seasonal_temp_mean = 23.2,   # Cooler than AR
    seasonal_temp_amplitude = 9.0,
    noise_std = 2.8
  )
)

sowing_dates <- list(
  "Early (Apr 15)" = 105,
  "Late (Jun 15)" = 166
)

## ── Parameters ──────────────────────────────────────────────
set.seed(42)
growing_season_days <- 130

cat("Generating realistic daily temperature and photoperiod data...\n")
cat("========================================================================\n\n")

all_data <- NULL

for (loc_idx in seq_len(nrow(locations))) {
  loc <- locations[loc_idx, ]
  params <- location_params[[loc$location]]

  for (sow_idx in seq_along(sowing_dates)) {
    sow_name <- names(sowing_dates)[sow_idx]
    sow_doy <- sowing_dates[[sow_idx]]

    cat(sprintf("%s - %s\n", loc$location, sow_name))

    # Create sequence of dates
    start_date <- as.Date(paste0("2023-", sow_doy), "%Y-%j")
    date_seq <- seq(start_date, by = "1 day", length.out = growing_season_days)

    scenario_data <- tibble(
      day_in_season = seq_len(growing_season_days),
      calendar_date = date_seq,
      calendar_doy = lubridate::yday(date_seq),
      location = loc$location,
      latitude = loc$lat,
      sowing_date = sow_name,
      sowing_doy = sow_doy,
      scenario = paste0(loc$location, " - ", sow_name)
    ) %>%
      mutate(
        # Temperature model: seasonal curve + daily noise
        season_phase = pmin(pmax((calendar_doy - 105) / 180 * pi, 0), pi),
        seasonal_mean = params$seasonal_temp_mean +
                       params$seasonal_temp_amplitude * sin(season_phase),
        daily_noise = rnorm(n(), mean = 0, sd = params$noise_std),
        temp_mean = seasonal_mean + daily_noise,
        # Diurnal temperature range ~9°C
        temp_min = temp_mean - 4.5,
        temp_max = temp_mean + 4.5,
        # Calculate photoperiod
        photoperiod_hours = calc_photoperiod(calendar_doy, latitude),
        # Round for output
        temp_mean = round(temp_mean, 1),
        temp_min = round(temp_min, 1),
        temp_max = round(temp_max, 1),
        photoperiod_hours = round(photoperiod_hours, 2)
      ) %>%
      select(location, latitude, sowing_date, sowing_doy,
             calendar_date, calendar_doy, day_in_season,
             temp_mean, temp_min, temp_max, photoperiod_hours, scenario)

    all_data <- bind_rows(all_data, scenario_data)
    cat(sprintf("  ✓ Generated %d days of data\n\n", nrow(scenario_data)))
  }
}

## ── Save to CSV ──────────────────────────────────────────────
output_file <- "temperature_and_photoperiod_data.csv"
write.csv(all_data, output_file, row.names = FALSE)

cat("========================================================================\n")
cat(sprintf("✓ Data saved to: %s\n", output_file))
cat(sprintf("✓ Total records: %d\n", nrow(all_data)))

## ── Summary statistics ──────────────────────────────────────
cat("\nTemperature range across all scenarios:\n")
temps <- all_data$temp_mean
cat(sprintf("  Min: %.1f°C\n", min(temps, na.rm = TRUE)))
cat(sprintf("  Max: %.1f°C\n", max(temps, na.rm = TRUE)))
cat(sprintf("  Mean: %.1f°C\n\n", mean(temps, na.rm = TRUE)))

cat("Photoperiod range across all scenarios:\n")
pps <- all_data$photoperiod_hours
cat(sprintf("  Min: %.1f hours\n", min(pps, na.rm = TRUE)))
cat(sprintf("  Max: %.1f hours\n\n", max(pps, na.rm = TRUE)))

# By scenario
cat("Temperature summary by scenario:\n")
summary_by_scenario <- all_data %>%
  group_by(scenario) %>%
  summarise(
    temp_mean = round(mean(temp_mean, na.rm = TRUE), 1),
    temp_range = paste0(round(min(temp_mean, na.rm = TRUE), 1), "–",
                        round(max(temp_mean, na.rm = TRUE), 1)),
    .groups = "drop"
  )
print(summary_by_scenario)

cat("\nPhotoperiod summary by scenario:\n")
summary_by_pp <- all_data %>%
  group_by(scenario) %>%
  summarise(
    pp_range = paste0(round(min(photoperiod_hours, na.rm = TRUE), 1), "–",
                      round(max(photoperiod_hours, na.rm = TRUE), 1)),
    .groups = "drop"
  )
print(summary_by_pp)

cat("\n========================================================================\n")
cat("NOTE: This is climate-realistic simulated data for visualization.\n")
cat("      For actual NASA POWER data, use: fetch_nasa_power_2010_2025.R\n")
